src/utils/include/utils/Array.hpp

```diff
@@ -125,6 +125,42 @@ template <typename T, std::size_t N> struct Array {
     return ret;
   }
 
+#ifdef __HCC__
+  __attribute__((annotate("serialize"))) void
+  __cxxamp_serialize(Kalmar::Serialize &s) const {
+    for (int i = 0; i < N; ++i)
+      s.Append(sizeof(T), &m_storage[i]);
+    for (int i = N; i < 10; ++i)
+      s.Append(sizeof(T), T());
+  }
+
+  __attribute__((annotate("user_deserialize"))) void
+  Foo(T x0, T x1, T x2, T x3, T x4, T x5, T x6, T x7, T x8, T x9)[[cpu]][[hc]] {
```

Comment:
What is the function of this method, and why is it called `Foo`?

Reply:
This is the deserialization method and it needs as many arguments as there were serialized parameters. `Foo` is a copy-and-paste leftover (it should actually be the name of the class, to be another constructor). It seems like it is resolved via the annotation attribute, so the name does not matter.

I think I may have found a simpler solution, will update the pull request soon.